libr/util/str.c

```diff
@@ -10,6 +10,76 @@
 #include <ctype.h>
 #include <stdarg.h>
 
+R_API RCharset *r_charset_new(size_t size) {
+	RCharset* rcs = (RCharset*) malloc (sizeof(RCharset));
+	rcs->custom_charset = NULL;
+	rcs->size = size;
+	rcs->remaining = 0;
+
+	return rcs;
+}
+
+R_API RCharsetRune *r_charset_rune_new(const ut8 *ch, const ut8 *hx) {
+	RCharsetRune* rcr = (RCharsetRune*) malloc (sizeof(RCharsetRune));
+	rcr->ch = (ut8 *) malloc (sizeof(rcr->ch));
+	strcpy ((char *) rcr->ch, (char *) ch);
+	rcr->hx = (ut8 *) malloc (sizeof(rcr->hx));
+	strcpy ((char *) rcr->hx, (char *) hx);
+	rcr->left = NULL;
+	rcr->right = NULL;
+
+	return rcr;
+}
+
+R_API void r_charset_rune_free(RCharsetRune *rcr) {
+	free (rcr->ch);
+	free (rcr->hx);
+	free (rcr);
+}
+
+R_API RCharsetRune *add_rune(RCharsetRune *rcsr, const ut8 *ch, const ut8 *hx) {
+	if (rcsr == NULL) {
+		rcsr = r_charset_rune_new(ch, hx);
+	}
+
+	if (strcmp ((char *)hx, (char *)rcsr->hx) < 0) {
+		rcsr->left = add_rune(rcsr->left, ch, hx);
+	} else if (strcmp ((char *)hx, (char *)rcsr->hx) > 0) {
+		rcsr->right = add_rune(rcsr->right, ch, hx);
+	} else {
+        if(strcmp ((char *)ch, (char *)rcsr->ch) > 0) {
+            rcsr->left = add_rune(rcsr->left, ch, hx);
+        }
+        else if(strcmp ((char *)ch, (char *)rcsr->ch) < 0){
+            rcsr->right = add_rune(rcsr->right, ch, hx);
+        }
+    }
+
+	return rcsr;
+}
+
+R_API RCharsetRune *search_from_hex(RCharsetRune *rcsr, const ut8 *hx) {
+	if (rcsr == NULL)
+	    return NULL;
+	else if ( strcmp ((char *)rcsr->hx, (char *)hx) == 0)
+	    return rcsr;
+	else {
+		RCharsetRune *left = search_from_hex(rcsr->left, hx);
+	    return left? left:  search_from_hex(rcsr->right, hx);
+	}
+}
+
+R_API RCharsetRune *search_from_char(RCharsetRune *rcsr, const ut8 *ch) {
+	if (rcsr == NULL)
+	    return NULL;
+	else if ( strcmp ((char *)rcsr->ch, (char *)ch) == 0)
+	    return rcsr;
+	else {
+		RCharsetRune *left = search_from_char(rcsr->left, ch);
+	    return left? left:  search_from_char(rcsr->right, ch);
```

Comment:
not sure how that left/right runes thing work. can you show an example?

Reply:
though that was 1:1 conversions

Reply:
@trufae exactly! It used to be a 1 to 1 conversion. Hopefully, it is not a 1 to 1 conversion anymore. I read your pad. Now this command handles more conversions. Example:
`0x4a => <PKMN>`

It should also convert something like:
`0x0102 => a`
It is an example that I imagine. Not a real architecture. I have not tested the last example.